### DISCUSSION TITLE
packageRules[n].extends doesn't appear in the schema / doesn't seem to work

### DISCUSSION THREAD

**Discussion by poulet42** [renovatebot/renovate#29673] [Request Help] [answered]
### What would you like help with?

I think I found a bug

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

gitlab, 37.408.0

### Please tell us more about your question or problem

packageRules documentation states we can use extends inside packageRules to extend different configuration based on a directory, when I try to edit my configuration to match something like this, I don't get autocomplete for the "extends" key, and looking at the [$schema](https://docs.renovatebot.com/renovate-schema.json), it doesn't seem to be documented there either. In addition this doesn't seem to work at all, it seems only the root extends got used here.

<img width="719" alt="Screenshot 2024-06-14 at 11 21 55" src="https://github.com/renovatebot/renovate/assets/16135480/c8cec111-41f6-4644-b5f4-0bad25fb5b02">

### Logs (if relevant)

<details><summary>Logs</summary>

```

Replace this text with your logs, between the starting and ending triple backticks

```

</details>

**Comment by rarkins** [comment]
Unfortunately the schema is not perfect, and it's not the best investment in our time to make it perfect instead of making more features. There's also a built-in validator for Renovate you can read more about here (also not perfect): https://docs.renovatebot.com/config-validation/#config-validation

**Reply by poulet42** [reply]
Thanks for the quick answer 🙏 , tried to run the validator and it doesn't seem happy with nested "extends" (using the same notation as the doc screenshot above). For the context I'm trying to setup renovate in a monorepo (with apps/** and packages/** directories). 
If the configuration syntax above is not valid, how canI have config overrides based on directories/workspaces?

**Reply by rarkins** [reply]
You never _need_ to use extends within packageRules, so I advise you actively avoid it if you're having troubles with it. You would override things by having packageRules which repeat the same config which the preset has. You can see the contents of the preset in docs or source code of Renovate

**Reply by poulet42** [reply]
Sorry, I think I didn't get exactly your last answer; 🙇 

I have a mix of apps and libs in a monorepo. according to this article [in your docs](https://docs.renovatebot.com/dependency-pinning/#so-whats-best) I should apply different pinning strategies for apps and consumed libraries. How can I tell renovate to do so, when everything lives under one repo?

**Reply by rarkins** [reply, marked as answer]
Use matchFileNames patterns like you originally suggested, but avoid using extends - just use regular config (e.g. `semanticCommits`)

**Reply by poulet42** [reply]
Awesome! I will try this :) Thank you
